src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,6 +299,34 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn max-command-size
+  "Returns the max command size relative to the current max heap. This
+  number was reached through testing of large catalogs and 1/205 was
+  the largest catalog that could be processed without GC or out of
+  memory errors"
+  []
+  (-> (Runtime/getRuntime)
+      .maxMemory
+      (/ 205)))
+
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app]
+  (let [max-content-length (max-command-size)]
+    (fn [req]
+      (if (and
+           (= :post (:request-method req))
+           (> (request/content-length req)
+              max-content-length))
+        (do
+          (.close (:body req))
+          {:status 413
+           :headers {}
+           :body "POST body too large"})
```

Comment:
I'll put a warning in there, that's easy to do. The tough part is getting good information about the error (see below). I'd be interested in thoughts there as right now I don't have a solid ways to get that information. 

I talked with @ajroetker separately about this showing through in the terminus. It will as we're returning an error response and we have error handling/logging in the terminus to handle that side of it.